src/java/main/org/apache/zookeeper/server/DataTree.java

```diff
@@ -478,7 +478,10 @@ public void createNode(final String path, byte data[], List<ACL> acl,
                 HashSet<String> list = ephemerals.get(ephemeralOwner);
                 if (list == null) {
                     list = new HashSet<String>();
-                    ephemerals.put(ephemeralOwner, list);
+                    HashSet<String> _list;
+                    if ((_list = ephemerals.putIfAbsent(ephemeralOwner, list)) != null) {
+                        list = _list;
+                    }
                 }
                 synchronized (list) {
```

Comment:
Even if when `createNode` is called by multi-thread.this fix is also not thread-safe.
Step1:One thread has acquired the `list` lock and enters into L487,add a element into  `list`
Step2:But at the same time just after Step1, another thread enters into L483,the reference of `list` is pointed to the old `_list` object. It will cause thread-unsafe
BTW: **Breaking Bad** is the best U.S.TV series

Reply:
I'm not sure about that.

ephemerals.putIfAbsent() is thread safe: it guarantees that newly created instance will be added to the HashMap only if the key is associated with a null value. Otherwise you will get the instance already present in the HashMap.
Either case `list` will point to the one and only instance present in the hashmap.

In your example when Step2 enters L483 it will get back the instance which was created and added to the HashMap by Step1 before it acquired the `list` lock. And that instance cannot be replaced in the HashMap with this code.

After all `synchronized (list)` only guards access to the List, HashMap is guarded by putIfAbsent().
I still believe that `computeIfAbsent()` is more convenient and readable here.

Reply:
@anmolnar 
I will test it with that unit test , using a 'threadpool' and create 1000000 nodes to reproduce what I guess。
I wish I was wrong,but if i'm right ,you need to watch Breaking Bad series (smirk).